From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bjarne Koll <git@lynxplay.dev>
Date: Tue, 20 May 2025 01:36:29 +0200
Subject: [PATCH] Add preliminary position support

The paper introduced position type is currently highly experimental
however, ktp and its consuming plugins aim to use this API.
For this, this patch adds fundamental support to the API for position
usage.

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 7aaeb4854144d35ff45a412035d38b415f593c27..5b8c947df9adea00b4d192cdac804ee6cb71457d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -90,6 +90,7 @@ import org.bukkit.Note;
 import org.bukkit.Particle;
 import org.bukkit.Raid;
 import org.bukkit.Sound;
+import org.bukkit.SoundCategory;
 import org.bukkit.TreeType;
 import org.bukkit.World;
 import org.bukkit.WorldBorder;
@@ -1065,7 +1066,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
     @Override
     public Collection<Entity> getNearbyEntities(Location location, double x, double y, double z) {
-        return this.getNearbyEntities(location, x, y, z, null);
+        return this.getNearbyEntities(location, x, y, z, (Predicate<? super Entity>) null); // KTP - preliminary position support
     }
 
     @Override
@@ -2505,4 +2506,20 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     }
     // KTP end - expose low level steering interceptors
 
+    // KTP start - preliminary position support
+    @Override
+    public void playSound(@org.jetbrains.annotations.NotNull final io.papermc.paper.math.Position position,
+                          @org.jetbrains.annotations.NotNull final Sound sound,
+                          @org.jetbrains.annotations.NotNull final SoundCategory category,
+                          final float volume,
+                          final float pitch) {
+        if (position == null || sound == null || category == null) return;
+
+        double x = position.x();
+        double y = position.y();
+        double z = position.z();
+
+        this.getHandle().playSound(null, x, y, z, CraftSound.bukkitToMinecraft(sound), SoundSource.valueOf(category.name()), volume, pitch);
+    }
+    // KTP end - preliminary position support
 }
